jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/RabitTracker.scala

```diff
@@ -0,0 +1,158 @@
+  */
+class RabitTracker(numWorkers: Int, port: Option[Int] = None,
+                   maxPortTrials: Int = 1000,
+                   workerConnectionTimeout: Duration = 10 minutes) extends IRabitTracker {
+  import scala.collection.JavaConverters._
+
+  require(numWorkers >=1, "numWorkers must be greater than or equal to one (1).")
+
+  val system = ActorSystem.create("RabitTracker")
+  val handler = system.actorOf(RabitTrackerHandler.props(
+    numWorkers, maxPortTrials, workerConnectionTimeout), "Handler")
+  implicit val askTimeout: akka.util.Timeout = akka.util.Timeout(1 second)
+
+  val futureWorkerEnvs: Future[Map[String, String]] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestBoundFuture, 5 seconds)
+      .asInstanceOf[Future[Map[String, String]]]
+  ) match {
+    case Success(fut) => fut
+    case Failure(e) =>
+      val delayedFailure = Promise[Map[String, String]]()
+      delayedFailure.failure(e)
+      delayedFailure.future
+  }
+
+  // a future for XGBoost worker execution
+  val futureCompleted: Future[Int] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestCompletionFuture, 5 seconds)
+      .asInstanceOf[Future[Int]]
+  ) match {
+    case Success(fut) => fut
+    case Failure(_) =>
+      val delayedFailure = Promise[Int]()
+      // mark 0 worker as completed, will trigger AssertionError
+      // when call waitFor()
+      delayedFailure.success(0)
+      delayedFailure.future
+  }
+
+  /**
+    * Start the Rabit tracker.
+    * @param timeout timeout for the tracker to bind to an address.
+    * @return
+    */
+  def start(timeout: Duration = 30 seconds): Boolean = {
+    handler ? RabitTrackerHandler.StartTracker(
+      new InetSocketAddress(InetAddress.getLocalHost,
+        // randomly allocate an ephemeral port if port is not specified
+        port.getOrElse(new Random().nextInt(61000 - 32768) + 32768)
+      ))
+
+    Try(Await.ready(futureWorkerEnvs, timeout)).isSuccess
+  }
+
+  def start(): Boolean = start(30 seconds)
+  def start(timeout: Long, unit: TimeUnit): Boolean = start(Duration(timeout, unit))
+
+  def getWorkerEnvs: java.util.Map[String, String] = {
+    new java.util.HashMap((Await.result(futureWorkerEnvs, 0 nano) ++ Map(
+        "DMLC_NUM_WORKER" -> numWorkers.toString,
+        "DMLC_NUM_SERVER" -> "0"
+    )).asJava)
+  }
+
+  /**
+    * Wait for all workers to complete assigned tasks in a blocking fashion.
+    * @param atMost the maximum execution time for the workers. By default,
+    *     the tracker waits for the workers indefinitely.
+    * @return the number of completed workers.
+    */
+  // @throws[TimeoutException]
+  def waitFor(atMost: Duration = Duration.Inf): Int = {
+    // wait for all workers to complete synchronously.
+    Try(Await.result(futureCompleted, atMost)) match {
+      case Success(n) if n == numWorkers =>
+        system.shutdown()
+        0
```

Comment:
to obey with the best practice in Java, you may want to replace 0, 1, 2 with enums